File: Top_Interview_150/Array_String/cpp/122.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int maxProfit(vector<int> &prices)
    {
        int n = prices.size();
        vector<pair<int, int>> dp(n);
        dp[0] = {-prices[0], 0};

        for (int i = 1; i < n; ++i)
        {
            dp[i].first = max(dp[i - 1].second - prices[i], dp[i - 1].first);
            dp[i].second = max(dp[i - 1].first + prices[i], dp[i - 1].second);
        }

        return max(dp[n - 1].first, dp[n - 1].second);
    }
};

File: Top_Interview_150/Array_String/python/122.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp = [[0, 0] for _ in range(n)]
        dp[0][0] = -prices[0]

        for i in range(1, n):
            dp[i][0] = max(dp[i - 1][1] - prices[i], dp[i - 1][0])
            dp[i][1] = max(dp[i - 1][0] + prices[i], dp[i - 1][1])

        return max(dp[n - 1])


solution = Solution()
print(solution.maxProfit(prices=[7, 6, 4, 3, 1]))
